SE-1165

rajeeshknambiar opened the pull request:
Write EDI document processing progress to display in the UI, while processing is ongoing.

Commit:
[IMP] odoo_edi: Show 'Prepare'/'Execute' buttons only when processing not started

SE-1165

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

Commit:
[ADD] edi: Log edi document prepare/execute processing status

SE-1165

While a large EDI document is being prepared/executed, which may create
a large number of records taking long (in background processing etc.),
let support users view the progress logging messages.

As EDI documents are prepared/executed under a DB `savepoint`, the
progress log messages need to be committed to a related model
(`edi.document.progress`) in a separate cursor to not disturb the main
transcation that creates edi records and target records.

An odoo-core bug caused yet-to-be written edi/target records also
committed when the progress log is committed in a different cursor, it
is worked around by clearing the `self.env.all.towrite` dict (which
holds records to be committed) and `self.env.all.tocompute` (which
holds computed fields) before updating the progress record and
restoring it after. See Odoo upstream bug 65036.

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

Commit:
[ADD] edi: New group to view EDI document processing log

SE-1165

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

ws-mps commented on addons/edi/models/edi_document.py:
```diff
@@ -188,6 +190,22 @@ class EdiDocument(models.Model):
         copy=False,
         tracking=True,
     )
+    execution_state = fields.Selection(
```
This should be an EDI document type config. Where is this used?

ws-mps replied:
Probably needs renaming.

rajeeshknambiar replied:
May be `allow_process` itself?
Figuring out the exact location to use this flag (may be in `autocreate()`?)

ws-mps commented on addons/edi/models/edi_document.py:
```diff
@@ -188,6 +190,22 @@ class EdiDocument(models.Model):
+        help="* Disabled: only receive file, don't try to process automatically."\
+        "  Users will manually process it."\
+        "* Prepare: only prepare after receiving, users will manually execute it."\
+        "  This is can be used to check that the file is syntactically correct."\
```
Typo: This can be

ws-mps commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +410,54 @@ def commit_processing_state_change(self, processing_state=""):
+        encompassing transaction (EDI records processing) undisturbed."""
+        towrite = self.env.all.towrite.copy()
+        with self.pool.cursor() as temp_cr:
+            # NOTE: this method does a commit using another cursor, but it is called from inside a
```
The issue happens anyway with or without being inside a savepoint.

rajeeshknambiar replied:
True, this is just to give a background of the control flow.

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -7,6 +7,8 @@
 from odoo.tools.translate import _
 from ..tools import NoRecordValuesError
+from time import sleep
+from time import sleep
```
duplicate import

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -231,6 +249,10 @@ class EdiDocument(models.Model):
     output_count = fields.Integer(
         string="Output Count", compute="_compute_output_count", store=True
     )
+    #edi_progress = fields.One2many("edi.document.progress", "doc_id")
```
nit: these commented out fields needed?

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -249,10 +271,31 @@ def _compute_input_count(self):
+        for rec in self:
+            rec.executed_count = rec.edi_progress.executed_count
+
+    #@api.depends("edi_progress.progress_log")
```
nit: delete

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -249,10 +271,31 @@ def _compute_input_count(self):
         for doc in self:
             doc.output_count = len(doc.output_ids)
 
+    def _compute_prepared_count(self):
```
nit: maybe `_compute_prepared_count` and `_compute_progress_log` can be combined as they share same (no) dependency

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +410,54 @@ def commit_processing_state_change(self, processing_state=""):
+            edi_progress = EDIDocProgress.search([("doc_id", "=", temp_self.id)])
+            if not edi_progress:
+                edi_progress = EDIDocProgress.create({"doc_id": temp_self.id})
+            #old_count = edi_progress.executed_count
```
nit: delete

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +410,54 @@ def commit_processing_state_change(self, processing_state=""):
+                updated_log = "\n".join([edi_progress.progress_log or "", log_message])
+                try:
+                    edi_progress.write({"progress_log": updated_log})
+                    #temp_cr.commit()   # commit is automatically done on cursor close
```
nit: delete and move aside comment to `with`

Commit:
[ADD] edi: Processing granularity for received documents

SE-1165

Increase processing granularity when receiving an EDI documents so that
it can be only received, prepared or executed. EDI gateway has
`allow_process` setting that triggers `edi.transfer` with this flag.
Check the processing level defined on the document type to decided
whether the document should be prepared or processed automatically.

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

Commit:
[IMP] edi_sale: Log document processing progress in DB/UI

SE-1165

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

Commit:
[IMP] edi_stock: Log document processing progress in DB/UI

SE-1165

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

armandcela commented on addons/edi/models/edi_document.py:
```diff
@@ -124,7 +141,7 @@ def autocreate(self, inputs, allow_process=True):
                 "doc_type_id": autodetect.type.id,
             }
-            if allow_process:
+            if allow_process and self.processing_level in ("prepare", "execute"):
```
Is allow_process field used for other purposes, i thing a single field can handle what we want

ws-mps replied:
Allow process can come from the gateway, in case we want to generally disable the processing of files at the GW level instead of EDI document level. I think it is fine as it is. Thanks for the suggestion.

armandcela commented on addons/edi/models/edi_document.py:
```diff
@@ -249,10 +267,17 @@ def _compute_input_count(self):
         for doc in self:
             doc.output_count = len(doc.output_ids)
 
+    def _compute_progress_log(self):
```
Is it thought to have a many2one field on document doc_progress_id and progress_log a related field with store=False.
From the code it looks like there can be more than one edi doc progress for same edi document

ws-mps replied:
We should only have one EDI doc progress per EDI document.

armandcela replied:
I think we are handling on the code to make sure that there is only one edi document progress per edi, so is ok i believe like it is

ws-mps commented on addons/edi/models/edi_transfer.py:
```diff
@@ -175,20 +178,29 @@ def do_transfer(self, conn):
             self.send_outputs(conn)
 
-        _logger.info("%s transfer complete", self.gateway_id.name)
+        self.log_progress(doc, "%s transfer complete" %self.gateway_id.name)
```
This is about a whole transfer, no need to log any document progress, also there is no doc.

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +392,52 @@ def commit_processing_state_change(self, processing_state=""):
+            edi_progress = EDIDocProgress.search([("doc_id", "=", temp_self.id)])
+            if not edi_progress:
+                edi_progress = EDIDocProgress.create({"doc_id": temp_self.id})
+            if edi_progress:
```
nit: may be overly cautious check?

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +392,52 @@ def commit_processing_state_change(self, processing_state=""):
+
+    def reset_progress(self):
+        self.ensure_one()
+        EDIDocProgress = self.env["edi.document.progress"]
```
nit: This logic is repeated a bit, maybe we could make a helper `get_or_create_edi_progress(doc_id)`

PeterAlabaster replied:
I don't mind if this one is left, it's only repeated twice

rajeeshknambiar replied:
Yes; I was trying to avoid using `self` which has the main cursor inside `commit_execute_progress()` thus repeated the logic.

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +392,52 @@ def commit_processing_state_change(self, processing_state=""):
+        EDIDocProgress = self.env["edi.document.progress"]
+        edi_progress = EDIDocProgress.search([("doc_id", "=", self.id)])
+        if edi_progress:
+            #edi_progress.executed_count = 0
```
nit: remove

PeterAlabaster commented on addons/edi/models/edi_synchronizer.py:
```diff
@@ -352,19 +353,25 @@ def execute(self):
                 )
                 with self.statistics() as stats:
                     vals_list = [rec.target_values(rec._record_values()) for rec in batch]
+                    #_logger.info(f"%%%% Updating {vals_list}")
```
nit: remove, and remove other commented code further down

PeterAlabaster commented on addons/edi/models/edi_transfer.py:
```diff
@@ -175,20 +178,29 @@ def do_transfer(self, conn):
-                if prepared:
-                    _logger.info("%s executing %s", self.gateway_id.name, doc.name)
-                    executed = doc.action_execute()
+                if doc.doc_type_id.processing_level in ("prepare", "execute"):
```
Can users still manually process if the processing level is disabled?

rajeeshknambiar replied:
Yes. I believe the processing levels should only affect automatic processing by gateway.

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +392,52 @@ def commit_processing_state_change(self, processing_state=""):
+            EDIDocProgress = temp_self.env["edi.document.progress"]
+            edi_progress = EDIDocProgress.search([("doc_id", "=", temp_self.id)])
+            if not edi_progress:
+                edi_progress = EDIDocProgress.create({"doc_id": temp_self.id})
```
Is there a chance a user could have permission to manually execute an EDI document but not create EDIDocProgress records?

rajeeshknambiar replied:
Assuming that would not be the case, this model's permissions are also set using the same groups for `edi.document`.

PeterAlabaster commented on addons/edi_stock/models/edi_move_request_record.py:
```diff
@@ -165,3 +170,10 @@ def execute(self):
         cancel._action_cancel()
+
+    @api.model
+    def log_progress(self, doc, log_message):
```
May be worth putting _some_ of these on `edi.record` (`public/odoo_edi/addons/edi/models/edi_record.py`) so we don't need as many duplicate redefinitions.
I think we would still need one one for `EdiSaleRequestDocument` though

PeterAlabaster commented on addons/edi/models/edi_synchronizer.py:
```diff
@@ -375,19 +379,23 @@ def execute(self):
                                 # must be rolled back for the affected object.
                                 with self.env.cr.savepoint():
                                     rec[target].write(vals)
+                                    #rec.doc_id.commit_execute_progress()
```
nit: remove?

Commit:
[FIX] edi: Patch cursor handling for unit tests

SE-1165

Unit tests patched `commit` function for current cursor object to avoid
commits. For the progress reporting, this causes error as the
edi.document record/key is not found in the new cursor opened for
progress reporting. To avoid foreign key referencing, change the
`doc_id` field in edi.document.progress to an Integer field, yet keep
the link to the document manually.

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

ws-mps commented on addons/edi/models/edi_transfer.py:
```diff
@@ -138,11 +138,14 @@ def receive_inputs(self, conn):
             # Log created documents
             for doc in docs:
                 Audit.audit_attachments(self, doc.input_ids, body=(_("Created %s") % doc.name))
-                _logger.info(
```
There is no EDI progress here, this should be a normal log.

rajeeshknambiar commented on addons/edi/models/edi_document.py:
```diff
@@ -96,6 +96,23 @@ def _default_project_id(self):
+            ("disabled", "Disabled"),
+            ("prepare", "Prepare"),
+            ("execute", "Execute")
+        ],
```
The default should be set to `disabled`, I believe. The empty value is technical treated similar to `disabled` (i.e., no automatic prepare/execute).

rajeeshknambiar replied:
As clarified, default should be `executed`, code revised.

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -418,7 +420,11 @@ def commit_execute_progress(self, log_message):
                 edi_progress = EDIDocProgress.create({"doc_id": temp_self.id})
-            updated_log = "\n".join([edi_progress.progress_log or "", log_message])
+            # Update with the message, or reset it
+            if log_message:
```
nit: 
```python
updated_log = ""
if log_message:
    updated_log = "\n".join([edi_progress.progress_log or "", log_message]).lstrip()
```

rajeeshknambiar replied:
The variable is set in the `else` part too (to reset the logs), so it will be defined in either case. But I'll set it to be explicit.

PeterAlabaster replied:
We can drop the
```python
else:
    updated_log = ""
```

now

PeterAlabaster commented on addons/edi/models/edi_document.py:
```diff
@@ -427,14 +433,11 @@ def commit_execute_progress(self, log_message):
         # Restore the original records to write in the main transcation
         self.env.all.towrite = towrite
 
+    def commit_execute_progress(self, log_message):
```
Q: Is this function needed anymore? Or can we just call `create_or_update_edi_progress` directly?

Commit:
[IMP] edi: Delete any linked progress records when edi document is deleted

SE-1165

Signed-off-by: Rajeesh KV <rajeesh.kv@unipart.com>

kdwyer commented on addons/edi/models/edi_document.py:
```diff
@@ -367,6 +393,52 @@ def commit_processing_state_change(self, processing_state=""):
+            # the cursor. To work around, clear the pending writes & computes here; write the progress
+            # and commit; then restore those after.
+            temp_self = self.with_env(self.env(cr=temp_cr))
+            temp_self.env.all.towrite.clear()
```
Is there a race condition here? For example, what if some other cursor from the pool calls `commit()` while the dicts are empty? (I suppose the first commit after the dicts are restored would pick up the skipped updates, but ... :grimacing: 

Or have I missed something obvious due to lack of morning caffeine?

kdwyer replied:
It seems that dicts are shared across the request's env cursors rather than the whole pool, so I'll resolve this.
